public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -0,0 +1,260 @@
+
+        let frames = respByTarget.get(targIdent);
+
+        if (!frames) {
+          frames = [];
+          respByTarget.set(targIdent, frames);
+        }
+
+        frames.push(frame);
+      });
+
+      let outFrames: DataFrame[] = [];
+
+      respByTarget.forEach((respFrames, targIdent) => {
+        let cachedFrames = (targIdent ? this.cache.get(targIdent)?.frames : null) ?? [];
+
+        respFrames.forEach((respFrame: DataFrame) => {
+          // skip empty frames
+          if (respFrame.length === 0 || respFrame.fields.length === 0) {
+            return;
+          }
+
+          // frames are identified by their second (non-time) field's name + labels
+          // TODO: maybe also frame.meta.type?
+          let respFrameIdent = getFieldIdent(respFrame.fields[1]);
+
+          let cachedFrame = cachedFrames.find((cached) => getFieldIdent(cached.fields[1]) === respFrameIdent);
+
+          if (!cachedFrame) {
+            // append new unknown frames
+            cachedFrames.push(respFrame);
+          } else {
+            // we assume that fields cannot appear/disappear and will all exist in same order
+
+            // amend & re-cache
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            let prevTable: Table = cachedFrame.fields.map((field) => field.values.toArray()) as Table;
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+            let nextTable: Table = respFrame.fields.map((field) => field.values.toArray()) as Table;
+
+            let amendedTable = amendTable(prevTable, nextTable);
+
+            for (let i = 0; i < amendedTable.length; i++) {
+              cachedFrame.fields[i].values = new ArrayVector(amendedTable[i]);
+            }
+
+            cachedFrame.length = cachedFrame.fields[0].values.length;
+          }
+        });
+
+        // trim all frames to in-view range, evict those that end up with 0 length
+        let nonEmptyCachedFrames: DataFrame[] = [];
+
+        cachedFrames.forEach((frame) => {
+          // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+          let table: Table = frame.fields.map((field) => field.values.toArray()) as Table;
+
+          let trimmed = trimTable(table, newFrom, newTo);
+
+          if (trimmed[0].length > 0) {
+            for (let i = 0; i < trimmed.length; i++) {
+              frame.fields[i].values = new ArrayVector(trimmed[i]);
+            }
+            nonEmptyCachedFrames.push(frame);
+          }
+        });
+
+        this.cache.set(targIdent, {
+          sig: requestInfo.targSigs.get(targIdent)!,
+          frames: nonEmptyCachedFrames,
+          prevTo: newTo,
+        });
+
+        outFrames.push(...nonEmptyCachedFrames);
+      });
+
+      // transformV2 mutates field values for heatmap de-accum, and modifies field order, so we gotta clone here, for now :(
```

Comment:
Should we track this in an issue to make it better?

Reply:
it might be expensive to make copies continuously, so i'm not sure this is something that can/should be "fixed". but maybe.